src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs

```diff
@@ -2781,7 +2781,36 @@ internal static int LengthInBufferCells(string str, int offset, bool checkEscape
                 }
             }
 
-            return str.Length - offset - escapeSequenceAdjustment;
+            int length = 0;
+            foreach (char c in str)
+            {
+                length += LengthInBufferCells(c);
+            }
+
+            return length - offset - escapeSequenceAdjustment;
+        }
+
+        internal static int LengthInBufferCells(char c)
+        {
+            // The following is based on http://www.cl.cam.ac.uk/~mgk25/c/wcwidth.c
+            // which is derived from https://www.unicode.org/Public/UCD/latest/ucd/EastAsianWidth.txt
+
+            bool isWide = c >= 0x1100 &&
+                (c <= 0x115f || /* Hangul Jamo init. consonants */
+                 c == 0x2329 || c == 0x232a ||
+                 (c >= 0x2e80 && c <= 0xa4cf &&
+                  c != 0x303f) || /* CJK ... Yi */
+                 (c >= 0xac00 && c <= 0xd7a3) || /* Hangul Syllables */
+                 (c >= 0xf900 && c <= 0xfaff) || /* CJK Compatibility Ideographs */
+                 (c >= 0xfe10 && c <= 0xfe19) || /* Vertical forms */
+                 (c >= 0xfe30 && c <= 0xfe6f) || /* CJK Compatibility Forms */
+                 (c >= 0xff00 && c <= 0xff60) || /* Fullwidth Forms */
+                 (c >= 0xffe0 && c <= 0xffe6));
```

Comment:
For performance we could use a pattern like:
```
if ((uint)(c - 'a') <= (uint)('z' - 'a'))
```

Reply:
🤔 I wonder if the new `System.Text.Rune` type has some kind of provision for this sort of thing.

Reply:
@vexx32 I believe the Rune struct only helps with encoding and not with rendering

Reply:
@iSazonov unless it's proven to be faster, I think the current code is more understandable

Reply:
The pattern is used in Core in some places like  https://source.dot.net/#System.Private.CoreLib/TextInfo.cs,564
It is 3 operations vs 2 operations.